### PR TITLE
Deal with multiple sessions.

### PR DESCRIPTION
**Pull request by butwerenotthereyet**
Make tmux-easy-motion work in more than one session by:
- specifying the exact target (session, window, pane) of commands
- using a separate pipe for each session

Need to merge https://github.com/IngoMeyer441/tmux-easy-motion/pull/3 first.

### CODE REVIEW

**Comment by IngoMeyer441** [comment]
Thanks for your second PR. I need some more explanation: Did tmux-easy-motion not work for you in different sessions without targeting window and pane explicitly? When programming tmux-easy-motion, I thought it would be sufficient to target the active pane.

**Comment by butwerenotthereyet** [comment]
Thanks for looking.  Yeah, the PR tries to resolve some bugs I've seen while using tmux-easy-motion.  Here's an easy way to see one of these bugs on master:

- launch tmux with a minimal conf file `tmux -L test -f minimal.conf` such as
```
run-shell "/path/to/tmux-easy-motion/easy_motion.tmux"
```
- cat a large file
- create a new session
- cat a large file in that new session
- in a second terminal window (or via ssh) attach to the server again and switch to the second session
- in the first terminal window, switch to the first session
- in the first terminal window, start tmux-easymotion, for example with `prefix space s s`
- in the second terminal window, start tmux-easymotion, for example with `prefix space s a`
- in the second terminal window, press a key to choose a target
- at this point, the first session is broken as shown here:
Screen Shot 2021-01-18 at 10 41 23 AM


> I thought it would be sufficient to target the active pane.

Not having the session, pane and window ids which the client had active when launching easy_motion.sh available to work with results in having to do unnecessary work.  For example, using awk when you could just ask tmux for exactly what you want: https://github.com/IngoMeyer441/tmux-easy-motion/pull/6/files#diff-e015cc3ac522b202bb23852488a7ac2c019e2c82568ecb387b97b60ebbecdf4bL61 .  Another example, asking tmux for information that could have just been passed directly to the script: https://github.com/IngoMeyer441/tmux-easy-motion/pull/6/files#diff-e015cc3ac522b202bb23852488a7ac2c019e2c82568ecb387b97b60ebbecdf4bL52 .  Launching each of these subprocesses is slow, and launching many adds up.

I'll split the commit up into two parts, one which just changes the pipe, and the other which specifies tmux command targets, if the patch works without the second part (which I haven't tried).

**Comment by IngoMeyer441** [comment]
Did you really experience performance impacts? When I wrote the first version of tmux-easy-motion I did not really think about performance since I didn't see this plugin as a performance critical program. But it would of course be nice to minimize the delays between a motion command and the display of targets on screen.

**Comment by IngoMeyer441** [comment]
What I tried to say: If performance is not relevant, I think it is more important to have readable code and I first thought that passing many arguments would harm readability. But on a second look, the readability is not worse than my version and your changes save unneeded work, so great :+1:

**Comment by butwerenotthereyet** [comment]
Okay!  The `display-message -p` does make for readable code.

What the previous shell scripting for tmux I've done has indicated is that getting rid of subprocesses (`cut`, `awk`, `sed`, and friends) and subshell `$()` function calling can make a big difference.  And that's very noticeable when tmux isn't responsive as the script executes.  Here's a post chatting about how expensive subshelling is, for example: http://rus.har.mn/blog/2010-07-05/subshells/ .  And there are better ways of writing functions to avoid subshelling than the technique in that post.  You can use indirect references https://tldp.org/LDP/abs/html/ivr.html and take as an argument the name of the variable to assign.

**Comment by IngoMeyer441** [comment]
Indirect references in functions work with `eval` if I understand correctly. `eval` is also quite slow, isn't it?

**Comment by IngoMeyer441** [comment]
The great thing about this PR is that the `session-closed` hook can now be used to clean up the temporary session directories 👍.

**Comment by butwerenotthereyet** [comment]
Nice idea with `session-closed`!  I'll do that.

I used superfine to measure the performance of calling a string concatenation function 10000 times written 1. using subshelling 2. using an agreed variable name (`retval` in the post) and 3. using an indirect reference.  Subshelling was `8.619 s ± 0.153 s`.  Agreed variable name was `198.7 ms ± 14.7 ms`.  Indirect was `285.6 ms ± 5.9 ms`.  So, an indirect reference is a little slower than a named variable, and both are much faster than subshelling.

File: scripts/easy_motion.sh [hunk 1]
```diff
@@ -120,9 +140,20 @@ easy_motion_cleanup() {
...
+        session_id="${BASH_REMATCH[1]}"
+    fi
+
+    set -- "$1" "${session_id}" "${@:3}"
+
     read_options && \
-    easy_motion_setup && \
-    easy_motion "$@"
+    easy_motion_setup "${session_id}" "${window_id}" "${pane_id}" && \
+    easy_motion "$@" && \
```
Comment (IngoMeyer441) [inline scripts/easy_motion.sh:156]:
```suggestion
    easy_motion "$@"
```
The cleanup step after `easy_motion` should always be run.

File: scripts/pipe_target_key.sh [hunk 2]
```diff
@@ -5,41 +5,50 @@ SCRIPTS_DIR="${CURRENT_DIR}"
 
 # shellcheck source=./common_variables.sh
 source "${SCRIPTS_DIR}/common_variables.sh"
-
+# shellcheck source=./helpers.sh
+source "${SCRIPTS_DIR}/helpers.sh"
 
 create_target_key_pipe() {
```
Comment (IngoMeyer441) [inline scripts/pipe_target_key.sh:11]: This function is now absolete and can be deleted.

**Comment by IngoMeyer441** [comment]
> I used superfine to measure the performance of calling a string concatenation function 10000 times written 1. using subshelling 2. using an agreed variable name (`retval` in the post) and 3. using an indirect reference. Subshelling was `8.619 s ± 0.153 s`. Agreed variable name was `198.7 ms ± 14.7 ms`. Indirect was `285.6 ms ± 5.9 ms`. So, an indirect reference is a little slower than a named variable, and both are much faster than subshelling.

Ok, I learned something new, thanks :smile: 

I have only two minor change requests to the current code, the rest looks good to me!

File: scripts/pipe_target_key.sh (hunk 2 above)
Comment (butwerenotthereyet) [reply scripts/pipe_target_key.sh]: Fixed.

File: scripts/easy_motion.sh (hunk 1 above)
Comment (butwerenotthereyet) [reply scripts/easy_motion.sh]: Fixed.

**Comment by butwerenotthereyet** [comment]
Okay!  I made those changes and one more to take the `$` prefix from the session_id only when creating the path.  I think this is ready, and we can add the cleanup on session close in another commit.

**Comment by IngoMeyer441** [comment]
Thanks again for your time and work! :+1:

**Comment by IngoMeyer441** [comment]
Commit 524b40566888bdecd85834279f7c073aced9886e adds a `session-closed` hook to clean up the target key pipe temporary directories and it worked so far in my tests.

**Pull request merged**